interfaces/apparmor/backend.go

```diff
@@ -647,22 +629,6 @@ func addContent(securityTag string, snapInfo *snap.Info, cmdName string, opts in
 		policy = classicTemplate
 		ignoreSnippets = true
 	}
-	// When partial AppArmor is detected, use the classic template for now. We could
-	// use devmode, but that could generate confusing log entries for users running
-	// snaps on systems with partial AppArmor support.
-	if apparmor_sandbox.ProbedLevel() == apparmor_sandbox.Partial {
-		// By default, downgrade confinement to the classic template when
-		// partial AppArmor support is detected. We don't want to use strict
-		// in general yet because older versions of the kernel did not
-		// provide backwards compatible interpretation of confinement
-		// so the meaning of the template would change across kernel
-		// versions and we have not validated that the current template
-		// is operational on older kernels.
```

Comment:
My main concern about this PR is the comment from this snippet, do we know that _not downgrading_ snap confinement will result in working snaps across the set of supported systems? It feels like the answer is yes given that many AppArmor features have in fact been upstreamed, but I don't know for sure.

Said another way, do we understand the fallout from this change? Yes it will make more snaps more secure, but will it make other snaps not work at all on other distros like Debian which are not simply missing AppArmor patches in the kernel but in fact have incompatible AppArmor patches in the kernel which enforce policy in a way our policy does not reflect and thus break some snaps. 

Maybe it's okay that some snaps are broken if on the whole more snaps are secure, but I think we should attempt to understand that first.

Reply:
Can't say about Debian, but on Arch and Tumbleweed we haven't really observed any major hiccups so far, with just one exception. The incomplete fine grained Unix socket mediation broke accessing X11 on Tumbleweed after a change to `x11` interface landed, but I carry this patch https://github.com/bboozzoo/snapd/commit/2a5f97e61e1e6457bc7f813c7d56cbac00b39205 in in the downstream packaging. If what @anonymouse64 meant by incompatible patches is this functionality, then the Debian package should likely carry that patch too. Note that we do not downgrade the AppArmor snippets for interfaces, even if the base template was simplified.

Looking at the template, I would expect the following rules to be ineffective without the remaining non-upstreamed bits of AppArmor:

```
  unix (bind, listen) addr="@snap.@{SNAP_INSTANCE_NAME}.**",
  unix peer=(label=snap.@{SNAP_INSTANCE_NAME}.*),
...
  dbus (receive, send) peer=(label=snap.@{SNAP_INSTANCE_NAME}.*),
...
  dbus (send)
      bus={session,system}
      path=/org/freedesktop/DBus
      interface=org.freedesktop.DBus.Introspectable
      member=Introspect
      peer=(label=unconfined),
```

Where the unix rule probably allows all unix socket access, and dbus is ignored (would also require dbus to be built with apparmor support?).

I would say that the downstream packaging should just disable AppArmor during the build if there's any concern about compatibility, in which case they would need to carry this patch: #9773 (also carried in openSUSE) so that snapd does not do something silly in specific scenarios.

One thing we could probably do is to display a warning that confinement is partial, like we do for cgroups v2. This information is available to via `snap debug confinement` but it's not intuitive and maybe it's the reason for the Debian bug we have.

It would also be nice if I finally got down to implementing the AppArmor related enhancements in our spread tests. IMO we're finishing confinement tests too early after a simple check `"$(snap debug confinement)" != strict`, where we could check for a specific apparmor feature(s) and continue executing the test. I believe we would have a better view of what is actually supported.

Reply:
@anonymouse64 I understand the concern but I think we need to be realistic.

The strategy to wait for upstreaming of all apparmor bits failed.

Right now users have effectively no security and perhaps sometimes, broken snaps (like nextcloud which can reach to host's mysql socket) CC @kyrofa for context if required. This is a reality check that we need to change direction. This is that change.

Is it secure? No, but neither was the previous version - it had not apparmor at all.
Is it guaranteed to work? No, but neither was the previous version - due to lack of apparmor, apps might load or access content from the host and misbehave.

I think we just need to honestly look at what is the possible outcome from pursuing those distinct paths and see what is better for our users and what is more honest.

Reply:
For interested readers, @zyga is referring to [LP: #1819734](https://bugs.launchpad.net/snapd/+bug/1819734).

Because snaps bundle their dependencies, snap developers often find themselves integrating pieces of software into their snap where they're lacking in-depth familiarity. It's not unusual for these components to try and reach out to the system and then have reasonable fallback behavior then it learns that such a thing is impossible. It took me ages to figure out why mysql was failing on Debian, and then more ages to figure out a way to make it stop trying to grab that file.

Ultimately, I believe that if snapd wants a reasonable cross-distro story, it needs to take ownership of ensuring that snaps run similarly everywhere instead of leaving the onus on:

- The user, who has no idea that they ran `sudo snap install nextcloud` and now have a daemon running as unconfined root on their server until it's far too late because they are under the impression that snaps run the same everywhere
- The snap developer, who created a snap thinking "snaps run the same everywhere, I'll be able to package once and my users can use it wherever they want without increasing my support load" when in reality, non-Ubuntu snap users can easily become the majority of support requests because snaps run very differently on non-Ubuntu systems

Reply:
My understanding is that the reason for the current 'fail-open' approach was that we didn't want to appear to users to be offering strict confinement when in fact, due to missing support in the host's environment, various things would not actually be mediated properly (dbus etc) without this fallback. However, I agree with the reasoning in the comments of this PR that this approach has created a lot of difficulty in various situations, and obviously results in a less secure deployment.

As such this PR will create a more secure setup for some users as we will now enforce confinement in these cases - I agree this will likely break some snaps which now get denied things that may have previously worked by chance, and this will obviously have holes in the confinement if various features are not supported - but this should be a better overall result than the status quo where there is no confinement at all.

Reply:
We are currently exploring which if any snaps will be broken by this change with some help from the advocacy team. I think I am in general on-board with this change so long as we at least understand who/what we are breaking (though note the spread tests on debian here were all happy).